datafusion/optimizer/src/analyzer/mod.rs

```diff
@@ -164,18 +172,38 @@ impl Analyzer {
             log_plan(rule.name(), &new_plan);
             observer(&new_plan, rule.as_ref());
         }
-        // for easier display in explain output
-        check_plan(&new_plan).map_err(|e| {
+
+        // verify at the end, after the last LP analyzer pass.
+        assert_valid_semantic_plan(&new_plan).map_err(|e| {
             DataFusionError::Context("check_analyzed_plan".to_string(), Box::new(e))
         })?;
+
         log_plan("Final analyzed plan", &new_plan);
         debug!("Analyzer took {} ms", start_time.elapsed().as_millis());
         Ok(new_plan)
     }
 }
 
-/// Do necessary check and fail the invalid plan
-fn check_plan(plan: &LogicalPlan) -> Result<()> {
+/// These are invariants which should hold true before and after each analyzer.
+///
+/// This differs from [`LogicalPlan::assert_invariants`], which addresses if a singular
+/// LogicalPlan is valid. Instead this address if the analyzer (before and after)
+/// is valid based upon permitted changes.
+///
+/// Does not check elements which are mutated by the analyzers (e.g. the schema).
+fn assert_valid_semantic_plan(plan: &LogicalPlan) -> Result<()> {
+    plan.assert_invariants()?;
+
+    // TODO: should this be moved to LogicalPlan::assert_invariants?
+    assert_subqueries_are_valid(plan)?;
```

Comment:
The `assert_subqueries_are_valid()` runs [these checks](https://github.com/apache/datafusion/blob/bd2c975df88c9e95f2f6b5f2abed5638904e2ac0/datafusion/optimizer/src/analyzer/subquery.rs#L28-L34). The existing code only performs these checks during after all of the analyzer passes have completed. If I run these checks before the first analyzer pass => we get test failures.

My next steps will be to figure out which of these test failures are due to code bugs.

Reply:
Also note that what makes a valid plan (plan checks) are different from [invariants](https://datafusion.apache.org/contributor-guide/specification/invariants.html#id4). Which is why I have this plan check here (and not under `LogicalPlan::assert_invariants`). But I'm not sure.

Reply:
> If I run these checks before the first analyzer pass => we get test failures.

I think this is somewhat "expected" -- part of the Analyzer's job is to make LogicalPlan's more "valid" as described here https://docs.rs/datafusion/latest/datafusion/optimizer/trait.AnalyzerRule.html

Perhaps we can make this part of `LogicalPlan::assert_invariants`, with an enum that explains the rules being followed? 

```rust
enum InvariantLevel {
  /// Invariants that are always true in DataFusion `LogicalPlan`s
  /// such as the number of expected children and no duplicated output fields
  Always, 
  /// Invariants that must hold true for the plan to be "executable"
  /// such as the type and number of function arguments are correct and
  /// that wildcards have been expanded
  ///
  /// To ensure a LogicalPlan satisfies the `Executable` inariants, run the
  /// `Analyzer`
  Executable,
}

impl LogicalPlan {
  /// checks that the plan conforms to the listed invariant level, returning an Error if not
  fn check_invariants(&self, check: InvariantLevel) -> Result<()> {
  ...
  }
}
```

That way the `assert_subqueries_are_valid` check would always be available